### ISSUE TITLE
Error: Unexpected null value on `updateDataSource`

### ISSUE THREAD

**Issue by anuj9196** [syncfusion/flutter-widgets#1134]
Want to update data in the line chart, dynamically. The data is updated in the parent widget, so have added the following code to refresh the chart every 1 second
```dart
  @override
  void initState() {
    super.initState();

    _timer = Timer.periodic(const Duration(seconds: 1), (timer) {
      _voltageChartController?.updateDataSource(addedDataIndexes: [widget.voltageData!.length-1]);
      _currentChartController?.updateDataSource(addedDataIndexes: [widget.currentData!.length-1]);
    });
  }
```
<img width="880" alt="Screenshot 2023-03-17 at 8 24 22 AM" src="https://user-images.githubusercontent.com/6560085/225800664-e2f70d92-9aa7-4500-a907-fb822639c788.png">

But on running, it is giving error
```
Error: Unexpected null value.
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 266:49      throw_
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 561:63  nullCheck
packages/syncfusion_flutter_charts/src/chart/chart_series/series.dart 1321:59     [_addOrUpdateDataPoint]
packages/syncfusion_flutter_charts/src/chart/chart_series/series.dart 1307:7      [_addOrUpdateDataPoints]
packages/syncfusion_flutter_charts/src/chart/chart_series/series.dart 1288:7      updateDataSource
packages/electrogenics_controller/widgets/charts/line_chart.dart 63:32            <fn>
dart-sdk/lib/_internal/js_dev_runtime/private/isolate_helper.dart 80:17           <fn>
```

The data on the chart is displaying despite the error message. But it shows only when running 2nd time.

**Comment by Yuvaraj-Gajaraj** [comment]
We have tried to reproduce the reported issue that is getting null exceptions while updating the data dynamically using the [updateDataSource](https://pub.dev/documentation/syncfusion_flutter_charts/latest/charts/ChartSeriesController/updateDataSource.html) method as mentioned below code snippet, but the reported issue is not reproduced at our end, even though at the second time. So we kindly request you to share more information on the complete chart code snippet and screenshot or screen recordings. We have also shared the sample below for your reference, so please try to reproduce the reported issue in that sample and get back to us it will be more helpful to us assist you in a better way.



Code snippet:
```dart
@override
void initState() {
  Timer.periodic(const Duration(seconds: 1), (timer) {
  // Add data point to existing data source list
    chartData.add(ChartData(
        chartData.last.x + 1, Random().nextInt(100), Random().nextInt(100)));
    _chartSeriesController1
        .updateDataSource(addedDataIndexes: [chartData.length - 1]);
    _chartSeriesController2
        .updateDataSource(addedDataIndexes: [chartData.length - 1]);
  });
  super.initState();
}
```

Sample: [i448061.zip](https://github.com/syncfusion/flutter-widgets/files/11024786/i448061.zip)

**Comment by anuj9196** [comment]
@Yuvaraj-Gajaraj Sharing the complete code for your reference
```dart
import 'dart:async';
import 'dart:math';
...
import 'package:flutter/material.dart';
import 'package:syncfusion_flutter_gauges/gauges.dart';

class AnalyticsPage extends StatefulWidget {
  const AnalyticsPage({Key? key}) : super(key: key);

  @override
  State<AnalyticsPage> createState() => _AnalyticsPageState();
}

class _AnalyticsPageState extends State<AnalyticsPage> {
  final testDuration = 10; // Test duration in seconds
  int _count = 0;
  Timer? _timer;
  final _random = Random();

  final List<CurrentData> currentData = [];
  final List<VoltageData> voltageData = [];

  late double _liveCurrent = 0;
  late double _liveVoltage = 0;
  late String _testDurationProgress = '00:00:00';

  randomNumber({dValue = false}) {
    if (dValue) {
      return _random.nextDouble();
    }
    return _random.nextInt(100);
  }

  @override
  void dispose() {
    _timer?.cancel();
    super.dispose();
  }

  @override
  void initState() {
    super.initState();
    addHeaderToTable();
    _timer = Timer.periodic(const Duration(seconds: 1), (timer) {
      setState(() {
        _liveCurrent = randomNumber(dValue: true);
        _liveVoltage = randomNumber().toDouble();
        _testDurationProgress = intToTime(_count);

        currentData.add(CurrentData(_count.toString(), _liveCurrent));
        voltageData.add(VoltageData(_count.toString(), _liveVoltage.toInt()));

      _count++;
      // Cancel timer
      if (testDuration == _count) {
        setState(() {
          _testDurationProgress = 'Complete';
        });
        timer.cancel();
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: Navbar(
            title: Text('Testing'),
        ),
        body: Stack(
          children: [
            SingleChildScrollView(
              scrollDirection: Axis.vertical,
              child: Padding(
                padding: const EdgeInsets.all(10),
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    Card(
                      color: Colors.white,
                      child: Padding(
                        padding: const EdgeInsets.all(10),
                        child: Column(
                          children: [
                            const Text(
                              'Voltage/Current',
                              style: TextStyle(
                                  fontSize: 16, fontWeight: FontWeight.bold),
                            ),
                            LineChart(
                              voltageData: voltageData,
                              currentData: currentData,
                            )
                          ],
                        ),
                      ),
                    ),
                    Row(
                      children: [
                        Flexible(
                          flex: 2,
                          child: Card(
                            child: Padding(
                              padding: const EdgeInsets.all(10),
                              child: Gauge(
                                title: 'Voltage',
                                pointer: _liveVoltage,
                                minValue: 0,
                                maxValue: 150,
                                scales: [
                                  GaugeRange(
                                      startValue: 0,
                                      endValue: 50,
                                      color: Colors.green,
                                      startWidth: 10,
                                      endWidth: 10),
                                  GaugeRange(
                                      startValue: 50,
                                      endValue: 100,
                                      color: Colors.orange,
                                      startWidth: 10,
                                      endWidth: 10),
                                  GaugeRange(
                                      startValue: 100,
                                      endValue: 150,
                                      color: Colors.red,
                                      startWidth: 10,
                                      endWidth: 10)
                                ],
                              ),
                            ),
                          ),
                        ),
                        Flexible(
                          flex: 2,
                          child: Card(
                            child: Padding(
                              padding: const EdgeInsets.all(10),
                              child: Gauge(
                                title: 'Current',
                                pointer: _liveCurrent,
                                minValue: 0,
                                maxValue: 1,
                                scales: [
                                  GaugeRange(
                                      startValue: 0,
                                      endValue: 0.5,
                                      color: Colors.green,
                                      startWidth: 10,
                                      endWidth: 10),
                                  GaugeRange(
                                      startValue: 0.5,
                                      endValue: 1,
                                      color: Colors.orange,
                                      startWidth: 10,
                                      endWidth: 10),
                                ],
                              ),
                            ),
                          ),
                        ),
                      ],
                    ),
                  ],
                ),
              ),
            ),
          ],
        ));
  }

  /// Int to time convert
  String intToTime(int value) {
    int h, m, s;

    h = value ~/ 3600;
    m = ((value - h * 3600)) ~/ 60;
    s = value - (h * 3600) - (m * 60);

    String hourLeft = h.toString().padLeft(2, '0');
    String minuteLeft = m.toString().padLeft(2, '0');
    String secondsLeft = s.toString().padLeft(2, '0');

    return '$hourLeft:$minuteLeft:$secondsLeft';
  }
}
```

Chart widgets are
```dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:syncfusion_flutter_charts/charts.dart';

class VoltageData {
  final int voltage;
  final String time;

  VoltageData(this.time, this.voltage);
}

class CurrentData {
  final double current;
  final String time;

  CurrentData(this.time, this.current);
}

class LineChart extends StatefulWidget {
  final List<CurrentData>? currentData;
  final List<VoltageData>? voltageData;

  const LineChart(
      {Key? key, required this.currentData, required this.voltageData})
      : super(key: key);

  @override
  State<LineChart> createState() => _LineChartState();
}

class _LineChartState extends State<LineChart> {

  late ChartSeriesController? _voltageChartController;
  late ChartSeriesController? _currentChartController;

  Timer? _timer;

  @override
  void initState() {
    super.initState();

    _timer = Timer.periodic(const Duration(seconds: 1), (timer) {
      _voltageChartController?.updateDataSource(
          addedDataIndexes: [widget.voltageData!.length - 1]);
      _currentChartController?.updateDataSource(
          addedDataIndexes: [widget.currentData!.length - 1]);
    });
  }

  @override
  void dispose() {
    _timer?.cancel();
    widget.voltageData?.clear();
    widget.currentData?.clear();
    _voltageChartController = null;
    _currentChartController = null;

    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return SfCartesianChart(
      legend: Legend(
          isVisible: true, isResponsive: true, position: LegendPosition.bottom),
      enableAxisAnimation: true,
      primaryYAxis: NumericAxis(title: AxisTitle(text: 'volt')),
      // Initialize category axis
      primaryXAxis: CategoryAxis(title: AxisTitle(text: '(seconds)')),
      axes: [
        NumericAxis(
            name: 'yAxis', opposedPosition: true, title: AxisTitle(text: 'amp'))
      ],
      series: [
        LineSeries<VoltageData, String>(
          onRendererCreated: (ChartSeriesController controller) {
            _voltageChartController = controller;
          },
          dataSource: widget.voltageData!,
          xValueMapper: (VoltageData volt, _) => volt.time,
          yValueMapper: (VoltageData volt, _) => volt.voltage,
          animationDuration: 0,
          name: 'Voltage',
        ),
        LineSeries<CurrentData, String>(
            onRendererCreated: (ChartSeriesController controller) {
              _currentChartController = controller;
            },
            dataSource: widget.currentData!,
            xValueMapper: (CurrentData current, _) => current.time,
            yValueMapper: (CurrentData current, _) => current.current,
            animationDuration: 0,
            name: 'Current',
            yAxisName: 'yAxis')
      ],
    );
  }
}

//// Gauage widget
import 'package:electrogenics_controller/constants/Theme.dart';
import 'package:flutter/material.dart';
import 'package:syncfusion_flutter_gauges/gauges.dart';

class GaugeScale {
  final int startValue;
  final int endValue;
  final Color color;

  GaugeScale(
      {required this.startValue,
      required this.endValue,
      this.color = ArgonColors.success});
}

class Gauge extends StatefulWidget {
  final String? title;
  final List<GaugeRange> scales;
  final double pointer;
  final double minValue;
  final double maxValue;

  const Gauge(
      {Key? key,
      this.title,
      required this.scales,
      this.pointer = 0,
      this.minValue = 0,
      this.maxValue = 100})
      : super(key: key);

  @override
  State<Gauge> createState() => _GaugeState();
}

class _GaugeState extends State<Gauge> {
  late double _pointer = 0;

  @override
  void setState(VoidCallback fn) {
    _pointer = widget.pointer;
    super.setState(fn);
  }

  Widget _getRadialGauge() {
    return SfRadialGauge(
        title: GaugeTitle(
            text: widget.title!,
            textStyle:
                const TextStyle(fontSize: 20.0, fontWeight: FontWeight.bold)),
        axes: <RadialAxis>[
          RadialAxis(
              minimum: widget.minValue,
              maximum: widget.maxValue,
              ranges: widget.scales,
              pointers: <GaugePointer>[
                NeedlePointer(value: widget.pointer)
              ],
              annotations: <GaugeAnnotation>[
                GaugeAnnotation(
                    widget: Text(widget.pointer.toStringAsFixed(2),
                        style: const TextStyle(
                            fontSize: 25, fontWeight: FontWeight.bold)),
                    angle: 90,
                    positionFactor: 0.5)
              ])
        ]);
  }

  @override
  Widget build(BuildContext context) {
    return _getRadialGauge();
  }
}
```

**Comment by michelelacorte** [comment]
> We have tried to reproduce the reported issue that is getting null exceptions while updating the data dynamically using the [updateDataSource](https://pub.dev/documentation/syncfusion_flutter_charts/latest/charts/ChartSeriesController/updateDataSource.html) method as mentioned below code snippet, but the reported issue is not reproduced at our end, even though at the second time. So we kindly request you to share more information on the complete chart code snippet and screenshot or screen recordings. We have also shared the sample below for your reference, so please try to reproduce the reported issue in that sample and get back to us it will be more helpful to us assist you in a better way.
> 
> Code snippet:
> 
> ```dart
> @override
> void initState() {
>   Timer.periodic(const Duration(seconds: 1), (timer) {
>   // Add data point to existing data source list
>     chartData.add(ChartData(
>         chartData.last.x + 1, Random().nextInt(100), Random().nextInt(100)));
>     _chartSeriesController1
>         .updateDataSource(addedDataIndexes: [chartData.length - 1]);
>     _chartSeriesController2
>         .updateDataSource(addedDataIndexes: [chartData.length - 1]);
>   });
>   super.initState();
> }
> ```
> 
> Sample: [i448061.zip](https://github.com/syncfusion/flutter-widgets/files/11024786/i448061.zip)

I think the problem is when multiple indexes are passed to updateDataSource, I have the same problem...

**Comment by michelelacorte** [comment]
@anuj9196 I've created a PR to fix this (#1137), meanwhile you can manually edit cached library or import it from fork

**Comment by LokeshPalani** [comment]
The reported issue regarding While adding the data point using updateDataSource when having an empty data source CategoryAxis has been fixed and rolled out in our weekly patch release. To avoid this please upgrade chart package to the latest version below.
 
Version:  https://pub.dev/packages/syncfusion_flutter_charts/versions/21.1.39
